### ISSUE TITLE
Datastore bi-directional relationships cause infinite recursion

### ISSUE THREAD

**Issue by nickname110** [spring-cloud/spring-cloud-gcp#1389]
Is there any way to implement OneToMany or ManyToMany relationships in Google DataStore?

```
    @Entity
    public class Entity {
        @Id
        private Long id;
        @Reference
        private List<SubEntity> subEntity;
    
    
        public Building(List<SubEntity> subEntity) {
            this.subEntity = subEntity;
        }
    
        //getters and setters
    }
    
    @Entity
    public class SubEntity {
        @Id
        private Long id;
    
        @Reference
        private Entity entity;
    
        //getters and setters
    }
```

It's not working

**Comment by meltsufin** [comment]
There are two ways to model relationships with `@Descendants` and with `@Reference`.
Please see the [reference documentation](https://cloud.spring.io/spring-cloud-static/spring-cloud-gcp/1.1.0.RC2/single/spring-cloud-gcp.html#_ancestor_descendant_relationships).

You seem to be trying to use `@Reference`. What is the exact problem in your example? Since the relationship is bi-directional, you might be forgetting to set one side of the relationship.

**Comment by nickname110** [comment]
> There are two ways to model relationships with `@Descendants` and with `@Reference`.
> Please see the [reference documentation](https://cloud.spring.io/spring-cloud-static/spring-cloud-gcp/1.1.0.RC2/single/spring-cloud-gcp.html#_ancestor_descendant_relationships).
> 
> You seem to be trying to use `@Reference`. What is the exact problem in your example? Since the relationship is bi-directional, you might be forgetting to set one side of the relationship.


I'm trying to:
```
Entity entity = new Entity(Arrays.asList(subEntity));
entityRepository.save(entity);
subEntity.setEntity(entity);
subEntityRepository.save(subEntity);
```

And then:
```
java.lang.StackOverflowError: null
 ...(with big stacktrace)
```

**Comment by meltsufin** [comment]
OK, that a big help! The problem seems to be that the mapping is circular and the recursion of the save becomes infinite. It's definitely something we should address by keeping track of what is already saved while doing the recursion.
For now, I would recommend having the relationship unidirectional.

**Comment by meltsufin** [comment]
@nickname110 The fix has been released in version 1.1.1.
